packages/editor/src/lib/editor/shapes/shared/arrow/shared.ts

```diff
@@ -81,40 +87,113 @@ function getArrowTerminalInArrowSpace(
 }
 
 /** @public */
-export function getArrowTerminalsInArrowSpace(editor: Editor, shape: TLArrowShape) {
-	const arrowPageTransform = editor.getShapePageTransform(shape)!
-
-	let startBoundShapeId: TLShapeId | undefined
-	let endBoundShapeId: TLShapeId | undefined
+export interface TLArrowBindings {
+	start: TLArrowBinding | undefined
+	end: TLArrowBinding | undefined
+}
 
-	if (shape.props.start.type === 'binding' && shape.props.end.type === 'binding') {
-		startBoundShapeId = shape.props.start.boundShapeId
-		endBoundShapeId = shape.props.end.boundShapeId
+/** @public */
+export function getArrowBindings(editor: Editor, shape: TLArrowShape): TLArrowBindings {
+	const bindings = editor.getBindingsFromShape<TLArrowBinding>(shape, 'arrow')
+	return {
+		start: bindings.find((b) => b.props.terminal === 'start'),
+		end: bindings.find((b) => b.props.terminal === 'end'),
 	}
+}
 
-	const boundShapeRelationships = getBoundShapeRelationships(
-		editor,
-		startBoundShapeId,
-		endBoundShapeId
-	)
+/** @public */
+export function getArrowTerminalsInArrowSpace(
+	editor: Editor,
+	shape: TLArrowShape,
+	bindings: TLArrowBindings
+) {
+	const arrowPageTransform = editor.getShapePageTransform(shape)!
 
-	const start = getArrowTerminalInArrowSpace(
+	const boundShapeRelationships = getBoundShapeRelationships(
 		editor,
-		arrowPageTransform,
-		shape.props.start,
-		boundShapeRelationships === 'double-bound' || boundShapeRelationships === 'start-contains-end'
+		bindings.start?.toId,
+		bindings.end?.toId
 	)
 
-	const end = getArrowTerminalInArrowSpace(
-		editor,
-		arrowPageTransform,
-		shape.props.end,
-		boundShapeRelationships === 'double-bound' || boundShapeRelationships === 'end-contains-start'
-	)
+	const start = bindings.start
+		? getArrowTerminalInArrowSpace(
+				editor,
+				arrowPageTransform,
+				bindings.start,
+				boundShapeRelationships === 'double-bound' ||
+					boundShapeRelationships === 'start-contains-end'
+			)
+		: Vec.From(shape.props.start)
+
+	const end = bindings.end
+		? getArrowTerminalInArrowSpace(
+				editor,
+				arrowPageTransform,
+				bindings.end,
+				boundShapeRelationships === 'double-bound' ||
+					boundShapeRelationships === 'end-contains-start'
+			)
+		: Vec.From(shape.props.end)
 
 	return { start, end }
 }
 
+/**
+ * Create or update the arrow binding for a particular arrow terminal. Will clear up if needed.
+ * TODO(alex): #bindings find a better name for this
```

Comment:
`doTheArrowThingsThnx`

Reply:
but seriously after reading through the arrow util and stuff, i'd suggest `createOrUpdateArrowBinding` and `unbindArrowTerminal`